### PR TITLE
Driver events

### PR DESCRIPTION
**Pull request by ashwin-nat**
## 📝 Description

Added car damage, wing change, pitting race ctrl messages
Moved fastest lap from event to status message, since the event packet is not reliable
Fixed breakage in race control messages filtering
Added test for driver based race control message and propagation to session manager

---

## 📂 Type of change

- [ ] Bug fix 🐞
- [ ] New feature 🚀
- [ ] Refactor 🔨
- [ ] Documentation 📚
- [ ] Tests ✅
- [ ] Other: 

---

## 🧪 Backend Test Reports (required for Python/backend changes)

> ⚠️ If your PR includes backend Python changes, please fill out the sections below.

### ✅ Integration Tests

* [ ] All integration tests pass
* Attach test command/output:

```
<paste result here>
```

To run the integration test
```bash
poetry run python tests/integration_test/runner.py
```

---

## 🧱 `lib/` Directory Changes

* [ ] This PR modifies or adds files under `lib/`
* [ ] I have updated or added unit tests to reflect those changes

Explain your test updates here:

```
Explain what was added/updated, or write "N/A" if not applicable.
```

---

## 📋 General Checklist

* [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) guide
* [ ] My code follows project style conventions
* [ ] All new and existing tests pass
* [ ] I have added relevant documentation/comments
* [ ] I have reviewed my changes and believe they are ready to merge

---

### CODE REVIEW

**Comment by ashwin-nat** [comment]
@sourcery-ai review

File: lib/race_ctrl/driver_mgr.py [hunk 1]
```diff
@@ -40,15 +43,29 @@ class DriverRaceControlManager:
...
-        """Register a message reference if the driver is involved."""
+        Args:
+            message (RaceCtrlMsgBase): The message to register.
+            propagate (bool): Whether to propagate the message to the session manager.
+                            If true, the message will be added to the session manager as well.
+        """
         self.messages.append(message)
+        if propagate:
+            assert self.session_mgr
+            self.session_mgr.add_message(message, is_propagated=True)
```
Comment (SourceryAI) [inline lib/race_ctrl/driver_mgr.py:59]: **issue (bug_risk):** The use of 'assert self.session_mgr' may cause runtime errors if not registered.

If 'register_session_manager' is not called before 'add_message', an AssertionError will occur. Please ensure registration is guaranteed or handle this case more robustly.

**Review by SourceryAI — COMMENTED** [review]
Hey there - I've reviewed your changes - here's some feedback:

- You’ve defined DriverPittingRaceCtrlMsg (and other status messages) in both driver_event_messages.py and driver_status_messages.py—consolidate these into the new status module and remove the duplicate definitions.
- There’s a typo in the class name CarDamageRaceCtrlMerssage (extra “s” in Merssage)—please rename it to CarDamageRaceCtrlMessage for consistency.
- Renaming race_ctrl_msg_factory to race_ctrl_event_msg_factory is a breaking change; consider providing an alias or deprecating the old factory to avoid import errors in other parts of the codebase.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- You’ve defined DriverPittingRaceCtrlMsg (and other status messages) in both driver_event_messages.py and driver_status_messages.py—consolidate these into the new status module and remove the duplicate definitions.
- There’s a typo in the class name CarDamageRaceCtrlMerssage (extra “s” in Merssage)—please rename it to CarDamageRaceCtrlMessage for consistency.
- Renaming race_ctrl_msg_factory to race_ctrl_event_msg_factory is a breaking change; consider providing an alias or deprecating the old factory to avoid import errors in other parts of the codebase.

## Individual Comments

### Comment 1
<location> `apps/backend/state_mgmt_layer/data_per_driver/data_per_driver.py:1122-1131` </location>
<code_context>
+    def addCarDamageRaceCtrlMsg(self, car_damage: CarDamageData) -> None:
</code_context>

<issue_to_address>
**issue (typo):** The method name 'CarDamageRaceCtrlMerssage' contains a typo.

Please update the class name to 'CarDamageRaceCtrlMessage' for consistency.

Suggested implementation:

```python
    def addCarDamageRaceCtrlMessage(self, car_damage: CarDamageData) -> None:
        """Add race control messages for car damage changes

        Args:
            car_damage (CarDamageData): The car damage data
        """
        if not self.m_packet_copies.m_packet_car_damage:
            return

        changed_fields = self.m_packet_copies.m_packet_car_damage.diff_fields(car_damage,
                                                                        self.CAR_DMG_RACE_CTRL_MSG_INTERESTED_FIELDS)

```

If this method is called elsewhere in the codebase, you will also need to update those references to use the new method name `addCarDamageRaceCtrlMessage`.
</issue_to_address>

### Comment 2
<location> `lib/race_ctrl/messages/driver_status_messages.py:52` </location>
<code_context>
+            ret["driver-info"] = driver_info
+        return ret
+
+class CarDamageRaceCtrlMerssage(RaceCtrlMsgBase):
+    def __init__(self,
+                 timestamp: float,
</code_context>

<issue_to_address>
**issue (typo):** Typo in class name: 'CarDamageRaceCtrlMerssage' should be 'CarDamageRaceCtrlMessage'.

Please correct the spelling to maintain consistency.

```suggestion
class CarDamageRaceCtrlMessage(RaceCtrlMsgBase):
```
</issue_to_address>

### Comment 3
<location> `lib/race_ctrl/messages/driver_status_messages.py:106` </location>
<code_context>
+            ret["driver-info"] = driver_info
+        return ret
+
+class TyreChangeRaceControlMessage(RaceCtrlMsgBase):
+    def __init__(self,
+                 timestamp: float,
</code_context>

<issue_to_address>
**suggestion:** Class name 'TyreChangeRaceControlMessage' is inconsistent with other message class naming.

Recommend renaming to 'TyreChangeRaceCtrlMsg' to match existing class naming conventions.

Suggested implementation:

```python
class TyreChangeRaceCtrlMsg(RaceCtrlMsgBase):
    def __init__(self,
                 timestamp: float,
                 driver_index: int,
                 lap_number: int,
                 damaged_part: str,
                 old_value: Union[int, float],
                 new_value: Union[int, float]) -> None:
        """Driver pitting message

        Args:
            timestamp (float): Time at which the message was issued (seconds).
            driver_index (int): Index of the driver.

```

If there are any other references to `TyreChangeRaceControlMessage` elsewhere in the codebase (e.g., in imports, type hints, or instantiations), those should also be updated to `TyreChangeRaceCtrlMsg` for consistency.
</issue_to_address>

### Comment 4
<location> `lib/race_ctrl/driver_mgr.py:57-59` </location>
<code_context>
+                            If true, the message will be added to the session manager as well.
+        """
         self.messages.append(message)
+        if propagate:
+            assert self.session_mgr
+            self.session_mgr.add_message(message, is_propagated=True)

     def clear(self) -> None:
</code_context>

<issue_to_address>
**issue (bug_risk):** The use of 'assert self.session_mgr' may cause runtime errors if not registered.

If 'register_session_manager' is not called before 'add_message', an AssertionError will occur. Please ensure registration is guaranteed or handle this case more robustly.
</issue_to_address>

### Comment 5
<location> `tests/tests_race_ctrl.py:189-196` </location>
<code_context>
+            driver_index=1,
+            lap_number=10)
+        self.driver1_mgr.add_message(msg)
+        exported = self.session_mgr.toJSON(self.driver_info_dict)
+
+        self.assertEqual(exported[0]["id"], 0)
+        self.assertEqual(exported[0]["message-type"], "PITTING")
+        self.assertEqual(exported[0]["lap-number"], 10)
+        self.assertEqual(exported[0]["driver-info"]["name"], "Driver One")
+        self.assertEqual(exported[0]["driver-info"]["team"], "Alpha")
+        self.assertEqual(exported[0]["driver-info"]["driver-number"], 1)
+        self.assertEqual(len(self.driver1_mgr.messages), 1)
+        self.assertEqual(len(self.session_mgr.messages), 1)
</code_context>

<issue_to_address>
**suggestion (testing):** No tests for new race control message types: car damage, wing change, tyre change.

Add unit tests for CarDamageRaceCtrlMessage, WingChangeRaceCtrlMsg, and TyreChangeRaceControlMessage to cover creation, propagation, JSON export, and edge cases such as no damage, multiple changes per lap, and compound changes.

Suggested implementation:

```python
    async def test_driver_message(self):
        msg = DriverPittingRaceCtrlMsg(
            timestamp=9.0,
            driver_index=1,
            lap_number=10)
        self.driver1_mgr.add_message(msg)
        exported = self.session_mgr.toJSON(self.driver_info_dict)

        self.assertEqual(exported[0]["id"], 0)
        self.assertEqual(exported[0]["message-type"], "PITTING")
        self.assertEqual(exported[0]["lap-number"], 10)
        self.assertEqual(exported[0]["driver-info"]["name"], "Driver One")
        self.assertEqual(exported[0]["driver-info"]["team"], "Alpha")
        self.assertEqual(exported[0]["driver-info"]["driver-number"], 1)
        self.assertEqual(len(self.driver1_mgr.messages), 1)
        self.assertEqual(len(self.session_mgr.messages), 1)
        self.assertEqual(self.driver1_mgr.messages[0], self.session_mgr.messages[0])

    async def test_car_damage_race_ctrl_message(self):
        # No damage edge case
        msg = CarDamageRaceCtrlMessage(
            timestamp=12.0,
            driver_index=1,
            lap_number=11,
            damage_parts=[]
        )
        self.driver1_mgr.add_message(msg)
        exported = self.session_mgr.toJSON(self.driver_info_dict)
        self.assertEqual(exported[0]["message-type"], "CAR-DAMAGE")
        self.assertEqual(exported[0]["lap-number"], 11)
        self.assertEqual(exported[0]["damage-parts"], [])
        self.assertEqual(len(self.driver1_mgr.messages), 1)
        self.assertEqual(len(self.session_mgr.messages), 1)

        # Multiple damage parts
        msg2 = CarDamageRaceCtrlMessage(
            timestamp=13.0,
            driver_index=1,
            lap_number=11,
            damage_parts=["front wing", "rear wing"]
        )
        self.driver1_mgr.add_message(msg2)
        exported = self.session_mgr.toJSON(self.driver_info_dict)
        self.assertEqual(exported[1]["damage-parts"], ["front wing", "rear wing"])
        self.assertEqual(len(self.driver1_mgr.messages), 2)
        self.assertEqual(len(self.session_mgr.messages), 2)

    async def test_wing_change_race_ctrl_message(self):
        # Single wing change
        msg = WingChangeRaceCtrlMsg(
            timestamp=14.0,
            driver_index=1,
            lap_number=12,
            wing_type="front"
        )
        self.driver1_mgr.add_message(msg)
        exported = self.session_mgr.toJSON(self.driver_info_dict)
        self.assertEqual(exported[0]["message-type"], "WING-CHANGE")
        self.assertEqual(exported[0]["lap-number"], 12)
        self.assertEqual(exported[0]["wing-type"], "front")

        # Multiple wing changes in one lap
        msg2 = WingChangeRaceCtrlMsg(
            timestamp=14.5,
            driver_index=1,
            lap_number=12,
            wing_type="rear"
        )
        self.driver1_mgr.add_message(msg2)
        exported = self.session_mgr.toJSON(self.driver_info_dict)
        self.assertEqual(exported[1]["wing-type"], "rear")
        self.assertEqual(len(self.driver1_mgr.messages), 2)
        self.assertEqual(len(self.session_mgr.messages), 2)

    async def test_tyre_change_race_ctrl_message(self):
        # Single tyre change
        msg = TyreChangeRaceControlMessage(
            timestamp=15.0,
            driver_index=1,
            lap_number=13,
            tyre_type="soft"
        )
        self.driver1_mgr.add_message(msg)
        exported = self.session_mgr.toJSON(self.driver_info_dict)
        self.assertEqual(exported[0]["message-type"], "TYRE-CHANGE")
        self.assertEqual(exported[0]["lap-number"], 13)
        self.assertEqual(exported[0]["tyre-type"], "soft")

        # Compound change: tyre and wing in same lap
        msg2 = TyreChangeRaceControlMessage(
            timestamp=15.5,
            driver_index=1,
            lap_number=13,
            tyre_type="medium"
        )
        msg3 = WingChangeRaceCtrlMsg(
            timestamp=15.6,
            driver_index=1,
            lap_number=13,
            wing_type="front"
        )
        self.driver1_mgr.add_message(msg2)
        self.driver1_mgr.add_message(msg3)
        exported = self.session_mgr.toJSON(self.driver_info_dict)
        self.assertEqual(exported[1]["tyre-type"], "medium")
        self.assertEqual(exported[2]["wing-type"], "front")
        self.assertEqual(exported[1]["lap-number"], 13)
        self.assertEqual(exported[2]["lap-number"], 13)
        self.assertEqual(len(self.driver1_mgr.messages), 3)
        self.assertEqual(len(self.session_mgr.messages), 3)

```

- Ensure that CarDamageRaceCtrlMessage, WingChangeRaceCtrlMsg, and TyreChangeRaceControlMessage are imported at the top of the file.
- If your test setup uses a different manager or message propagation pattern, adjust the test code accordingly.
- If your toJSON method returns messages in a different order, update the index checks.
</issue_to_address>

### Comment 6
<location> `lib/race_ctrl/messages/driver_status_messages.py:31` </location>
<code_context>
                 ret["overtaken-info"] = overtaken_info
         return ret
+
+class DriverPittingRaceCtrlMsg(RaceCtrlMsgBase):
+    def __init__(self, timestamp: float, driver_index: int, lap_number: Optional[int] = None) -> None:
+        """Driver pitting message
</code_context>

<issue_to_address>
**issue (complexity):** Consider refactoring the four message classes to use a shared mix-in and move common logic to the base class to reduce boilerplate.

Here’s one way to collapse almost all of those four classes into one tiny constructor mix-in plus small subclasses, and lift your `driver-info` logic into `RaceCtrlMsgBase.toJSON()`.  This preserves 100% of your functionality but removes >70% of the boilerplate.

```python
# --- in base.py --------------------------------------------
class RaceCtrlMsgBase:
    def __init__(self,
                 timestamp: float,
                 message_type: MessageType,
                 involved_drivers: list[int],
                 lap_number: Optional[int] = None) -> None:
        …  # unchanged

    def toJSON(self, driver_info_dict: Optional[dict[int,Any]] = None) -> dict[str,Any]:
        base = {
            "timestamp":   self.timestamp,
            "message-type": self.message_type.value,
            "drivers":     self.involved_drivers,
            "lap":         self.lap_number,
            # … any other common fields
        }
        # move all your driver-info logic here
        if driver_info_dict and self.involved_drivers:
            info = driver_info_dict.get(self.involved_drivers[0])
            if info is not None:
                base["driver-info"] = info
        return base
```

```python
# --- in your new module -----------------------------------
from .base import MessageType, RaceCtrlMsgBase

class SimpleDriverMsg(RaceCtrlMsgBase):
    """common init for any single-driver control msg"""
    MESSAGE_TYPE: MessageType  # override in subclasses

    def __init__(self,
                 timestamp: float,
                 driver_index: int,
                 lap_number: Optional[int] = None) -> None:
        super().__init__(
            timestamp=timestamp,
            message_type=self.MESSAGE_TYPE,
            involved_drivers=[driver_index],
            lap_number=lap_number
        )

class DriverPittingRaceCtrlMsg(SimpleDriverMsg):
    MESSAGE_TYPE = MessageType.PITTING

class WingChangeRaceCtrlMsg(SimpleDriverMsg):
    MESSAGE_TYPE = MessageType.WING_CHANGE

class CarDamageRaceCtrlMsg(SimpleDriverMsg):
    MESSAGE_TYPE = MessageType.CAR_DAMAGE

    def __init__(self,
                 timestamp: float,
                 driver_index: int,
                 lap_number: int,
                 damaged_part: str,
                 old_value: Union[int, float],
                 new_value: Union[int, float]) -> None:
        super().__init__(timestamp, driver_index, lap_number)
        self.damaged_part = damaged_part
        self.old_value = old_value
        self.new_value = new_value

    def toJSON(self, driver_info_dict=None) -> dict[str, Any]:
        ret = super().toJSON(driver_info_dict)
        ret.update({
            "damaged-part": self.damaged_part,
            "old-value":    self.old_value,
            "new-value":    self.new_value,
        })
        return ret

class TyreChangeRaceCtrlMsg(SimpleDriverMsg):
    MESSAGE_TYPE = MessageType.TYRE_CHANGE

    def __init__(self,
                 timestamp: float,
                 driver_index: int,
                 lap_number: Optional[int],
                 old_tyre_compound: str,
                 old_tyre_index: int,
                 new_tyre_compound: str,
                 new_tyre_index: int) -> None:
        super().__init__(timestamp, driver_index, lap_number)
        self.old_tyre_compound = old_tyre_compound
        self.old_tyre_index    = old_tyre_index
        self.new_tyre_compound = new_tyre_compound
        self.new_tyre_index    = new_tyre_index

    def toJSON(self, driver_info_dict=None) -> dict[str, Any]:
        ret = super().toJSON(driver_info_dict)
        ret.update({
            "old-tyre-compound": self.old_tyre_compound,
            "old-tyre-index":    self.old_tyre_index,
            "new-tyre-compound": self.new_tyre_compound,
            "new-tyre-index":    self.new_tyre_index,
        })
        return ret
```

Benefits:

  * All the identical `super().__init__` calls live in `SimpleDriverMsg`
  * The one‐off payload fields only appear in the two classes that need them
  * The `driver-info` bit is implemented once on `RaceCtrlMsgBase.toJSON`
  * You preserve exactly the same public API and JSON output, but cut down ~75% of the boilerplate.
</issue_to_address>

### Comment 7
<location> `apps/backend/state_mgmt_layer/data_per_driver/data_per_driver.py:1154-1155` </location>
<code_context>
    def addCarDamageRaceCtrlMsg(self, car_damage: CarDamageData) -> None:
        """Add race control messages for car damage changes

        Args:
            car_damage (CarDamageData): The car damage data
        """
        if not self.m_packet_copies.m_packet_car_damage:
            return

        changed_fields = self.m_packet_copies.m_packet_car_damage.diff_fields(car_damage,
                                                                        self.CAR_DMG_RACE_CTRL_MSG_INTERESTED_FIELDS)
        for field, diff in changed_fields.items():
            new_value = diff["new_value"]
            old_value = diff["old_value"]
            if new_value > old_value:
                self.m_race_ctrl.add_message(CarDamageRaceCtrlMerssage(
                    timestamp=time.time(),
                    driver_index=self.m_index,
                    lap_number=self.m_lap_info.m_current_lap,
                    damaged_part=field,
                    old_value=old_value,
                    new_value=new_value
                ))
                msg_type_str = "car damage"
            else:
                self.m_race_ctrl.add_message(WingChangeRaceCtrlMsg(
                    timestamp=time.time(),
                    driver_index=self.m_index,
                    lap_number=self.m_lap_info.m_current_lap
                ))
                msg_type_str = "wing change"

            self.m_logger.debug("Driver %s - %s changed from %s to %s. Added %s race control message",
                                str(self), field, diff["old_value"], diff["new_value"], msg_type_str)
</code_context>

<issue_to_address>
**suggestion (code-quality):** Use previously assigned local variable [×2] ([`use-assigned-variable`](https://docs.sourcery.ai/Reference/Default-Rules/refactorings/use-assigned-variable/))

```suggestion
            self.m_logger.debug(
                "Driver %s - %s changed from %s to %s. Added %s race control message",
                str(self),
                field,
                old_value,
                new_value,
                msg_type_str,
            )
```
</issue_to_address>
~~~

</details>

***

Hi @ashwin-nat! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>
